File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/SmartStackRelevance.swift
import Foundation

// MARK: - SmartStackRelevance
//
// How much Forge is worth showing right now.
//
// The watch complications existed and never surfaced themselves. The Smart
// Stack ranks what it shows by the relevance each timeline entry declares, and
// every entry declared none — so Forge appeared only when the user scrolled to
// it, which is the opposite of what a proactive app is for. Meanwhile
// ContextEngine was already computing the moments that matter and discarding
// every one of them.
//
// Returns plain numbers rather than a TimelineEntryRelevance so the decision is
// testable without WidgetKit; the widgets extension does the one-line
// conversion.

public enum SmartStackRelevance {

    public struct Relevance: Equatable, Sendable {
        /// Higher wins a place further up the stack.
        public var score: Float
        /// How long the score holds. Nil means until the next timeline reload.
        public var duration: TimeInterval?

        public init(score: Float, duration: TimeInterval? = nil) {
            self.score = score
            self.duration = duration
        }
    }

    /// A workout in progress. Nothing outranks the thing happening now.
    public static let liveSession: Float = 100
    /// A moment the context engine actively flagged.
    public static let flagged: Float = 60
    /// Worth showing, no particular urgency.
    public static let ambient: Float = 20

    /// How close to the wind-down time counts as "now".
    public static let windDownWindowMinutes: Double = 90

    /// The scores are relative, not absolute. What matters is the ordering — a
    /// live workout above a wind-down prompt above an ambient glance — so that
    /// nothing can push a running session down the stack.
    public static func score(for snapshot: WatchSnapshot?, now: Date = Date()) -> Relevance? {
        guard let snapshot else { return nil }

        if let workout = snapshot.activeWorkout, workout.phase != .ended {
            // No duration: the app reloads timelines on every phase change, so
            // the score is replaced rather than expiring on a guess about how
            // long somebody trains.
            return Relevance(score: liveSession)
        }

        if let windDown = snapshot.tonightWindDown,
           isWithin(windDown, of: now, minutes: windDownWindowMinutes) {
            return Relevance(score: flagged, duration: windDownWindowMinutes * 60)
        }

        // A recommendation exists whenever the suggestion engine has something
        // to say — which, after a 90-minute desk block, is the reset it wants
        // to offer. Below wind-down because a reset keeps; a bedtime does not.
        if snapshot.recommendedPractice != nil {
            return Relevance(score: flagged * 0.7)
        }

        return Relevance(score: ambient)
    }

    /// Symmetric on purpose: the half hour after the wind-down window opened is
    /// at least as relevant as the half hour before it.
    private static func isWithin(_ date: Date, of now: Date, minutes: Double) -> Bool {
        abs(date.timeIntervalSince(now)) <= minutes * 60
    }
}

File: backend/infra/lambda/aria_core/smart_stack_relevance.py
"""How much Forge is worth showing in a glance / Smart Stack.

Python port of ForgeCore's ``SmartStackRelevance.swift``. WidgetKit conversion
stays on the client; this returns the numbers.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

LIVE_SESSION = 100.0
FLAGGED = 60.0
AMBIENT = 20.0
WIND_DOWN_WINDOW_MINUTES = 90.0


@dataclass(frozen=True)
class Relevance:
    score: float
    duration_seconds: float | None = None

    def to_dict(self) -> dict:
        payload = {"score": self.score}
        if self.duration_seconds is not None:
            payload["durationSeconds"] = self.duration_seconds
        return payload


def score(
    *,
    active_workout_phase: str | None = None,
    tonight_wind_down: datetime | None = None,
    recommended_practice: str | None = None,
    now: datetime | None = None,
) -> Relevance | None:
    """``active_workout_phase`` None or ``ended`` means no live session."""
    if now is None:
        now = datetime.now()

    if active_workout_phase is not None and active_workout_phase != "ended":
        return Relevance(score=LIVE_SESSION)

    if tonight_wind_down is not None:
        delta = abs((tonight_wind_down - now).total_seconds())
        if delta <= WIND_DOWN_WINDOW_MINUTES * 60:
            return Relevance(score=FLAGGED, duration_seconds=WIND_DOWN_WINDOW_MINUTES * 60)

    if recommended_practice is not None:
        return Relevance(score=FLAGGED * 0.7)

    return Relevance(score=AMBIENT)
